msal/src/test/java/com/microsoft/identity/client/e2e/tests/mocked/DeviceCodeFlowApiTest.java

```diff
@@ -0,0 +1,329 @@
+        final MicrosoftStsAuthorizationRequest authorizationRequest = mBuilder.setScope(null).build();
+        final AuthorizationResult authorizationResult = mStrategy.getDeviceCode(authorizationRequest, mUrlBody);
+        final MicrosoftStsAuthorizationErrorResponse authorizationErrorResponse = (MicrosoftStsAuthorizationErrorResponse) authorizationResult.getAuthorizationErrorResponse();
+
+        Assert.assertFalse(authorizationResult.getSuccess());
+        Assert.assertNull(authorizationResult.getAuthorizationResponse());
+
+        Assert.assertNotNull(authorizationErrorResponse);
+        Assert.assertEquals(ErrorStrings.INVALID_REQUEST, authorizationErrorResponse.getError());
+    }
+
+    @Test
+    public void testGetDeviceCodeFailureBadScope() throws IOException {
+        final MicrosoftStsAuthorizationRequest authorizationRequest = mBuilder.setScope("/").build();
+        final AuthorizationResult authorizationResult = mStrategy.getDeviceCode(authorizationRequest, mUrlBody);
+        final MicrosoftStsAuthorizationErrorResponse authorizationErrorResponse = (MicrosoftStsAuthorizationErrorResponse) authorizationResult.getAuthorizationErrorResponse();
+
+        Assert.assertFalse(authorizationResult.getSuccess());
+        Assert.assertNull(authorizationResult.getAuthorizationResponse());
+
+        Assert.assertNotNull(authorizationErrorResponse);
+        Assert.assertEquals(ErrorStrings.INVALID_SCOPE, authorizationErrorResponse.getError());
+    }
+
+    //===========================================================================================================
+    // Token Request Testing
+    //===========================================================================================================x
+    @Test
+    public void testDeviceCodeFlowTokenInvalidRequest() throws IOException, ClientException {
+        final MicrosoftStsOAuth2Configuration config = new MicrosoftStsOAuth2Configuration();
+        config.setAuthorityUrl(new URL(mUrlBody));
+        final OAuth2StrategyParameters options = new OAuth2StrategyParameters();
+        final OAuth2Strategy strategy = new MicrosoftStsOAuth2Strategy(config, options);
+
+        final TokenResult tokenResult = strategy.requestToken(mTokenRequest);
+        Assert.assertNull(tokenResult.getTokenResponse());
+        Assert.assertNotNull(tokenResult.getErrorResponse());
+        Assert.assertEquals(ErrorStrings.INVALID_REQUEST, tokenResult.getErrorResponse().getError());
+    }
+
+    @Test
+    public void testDeviceCodeFlowTokenExpiredToken() throws IOException, ClientException {
+        final MicrosoftStsOAuth2Configuration config = new MicrosoftStsOAuth2Configuration();
+        config.setAuthorityUrl(new URL(mUrlBody));
+        final OAuth2StrategyParameters options = new OAuth2StrategyParameters();
+        final OAuth2Strategy strategy = new MicrosoftStsOAuth2Strategy(config, options);
+
+        // Previously authenticated code
+        mTokenRequest.setDeviceCode(
```

Comment:
given that it's the same mTokenRequest object with the test case above, why won't I hit ErrorStrings.INVALID_REQUEST here?

Reply:
INVALID_REQUEST was triggered in the test case above this one because the device code was not included. This one has a device code that I already authenticated on so it returns expired token.